src/vs/editor/browser/viewParts/glyphMargin/glyphMargin.ts

```diff
@@ -183,7 +243,48 @@ export class GlyphMarginOverlay extends DedupOverlay {
 	}
 
 	// --- end event handlers
+	// --- begin widget management
+	public addWidget(widget: IGlyphMarginWidget): void {
+		const domNode = createFastDomNode(widget.getDomNode());
+
+		this._widgets[widget.getId()] = {
+			widget: widget,
+			preference: widget.getPosition(),
+			domNode: domNode
+		};
 
+		domNode.setPosition('absolute');
+		domNode.setAttribute('widgetId', widget.getId());
+		this.domNode.appendChild(domNode);
+
+		this.setShouldRender();
+	}
+
+	public setWidgetPosition(widget: IGlyphMarginWidget, preference: IGlyphMarginWidgetPosition): boolean {
+		const myWidget = this._widgets[widget.getId()];
+		if (myWidget.preference === preference) {
+			return false;
+		}
```

Comment:
I think the preferences should be compared structurally because they're objects, not primitives.